### PR TITLE
[BEAM-9167] Metrics extraction refactoring.

### PR DESCRIPTION
**Pull request by lostluck**
Changes up how metrics are accessed. Instead of having them globally available, they're exclusively available on a per-bundle metrics store, which reduces the overhead of creating metrics proxies and updating. 

In particular, this change moves the extraction of metrics to the FnAPI protos to the exec package, so the metrics package becomes proto agnostic. It's also forward looking, in that the metric extractors are typed with int64 to distinguish them for eventual inclusion of the float64 variants.

Adds additional benchmarking around specific parts of the work of modifying metrics. Benchmark values, as always, are relative to machine, OS, and Go version.

Two areas for subsequent work:
Left as an open question are parts are how users can get access to the store (or provide an extractor) for metrics after a job. Missing for this is an equivalent "ingestor" for accepting metrics from an outside source for use by runners.

For a more near term CL, is actually using the v2 metrics MonitoringInfo protos instead of the legacy style. This refactoring makes it easier to support both, without the metrics package needing to be tightly coupled to the protos.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lostluck** [comment]
R: @youngoli

**Comment by lostluck** [comment]
Run Go PostCommit

**Comment by lostluck** [comment]
Run Go PostCommit

File: sdks/go/pkg/beam/runners/direct/direct.go [hunk 1]
```diff
@@ -72,7 +72,9 @@ func Execute(ctx context.Context, p *beam.Pipeline) error {
 	if err = plan.Down(ctx); err != nil {
 		return err
 	}
-	metrics.DumpToLog(ctx)
+	// TODO(lostluck) 2020/01/24: What's the right way to expose the
+	// metrics store for the direct runner?
+	metrics.DumpToLogFromStore(ctx, plan.Store)
```
Comment (youngoli) [inline sdks/go/pkg/beam/runners/direct/direct.go]: To answer the TODO's question: This seems fine to me. What are the concerns you have about exposing the metrics like this?

File: sdks/go/pkg/beam/core/metrics/dumper.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		},
+		GaugeInt64: func(l Labels, v int64, t time.Time) {
+			m[l] = &gauge{v: v, t: t}
+		},
+	}
+	e.ExtractFrom(store)
+	dumpTo(m, p)
+}
+
+type metricDumper struct {
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/core/metrics/dumper.go]: Good catch. Oversight from a previous prototype.

File: sdks/go/pkg/beam/core/metrics/metrics.go [hunk 3]
```diff
@@ -144,63 +119,55 @@ func (ctx *beamCtx) String() string {
...
-		bctx.bs.css = append(bctx.bs.css, cs)
-		bctx.cs = cs
-		bctx.bs.mu.Unlock()
-		return cs
+	if set := ctx.Value(counterSetKey); set != nil {
+		return set.(*ptCounterSet)
 	}
-	panic("counterSet missing, beam isn't set up properly.")
-	return nil // never runs.
+	// This isn't a beam context, so we can't store the metric.
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/core/metrics/metrics.go]: The counterset is what's used to store the metric for bundle access, so it's accurate, but you're right that it's confusing. Rewording.

File: sdks/go/pkg/beam/runners/direct/direct.go (hunk 1 above)
Comment (lostluck) [reply sdks/go/pkg/beam/runners/direct/direct.go]: It's more of a general comment of how do we give users programmatic access to the metrics after pipeline completion. So there's nothing wrong with the way the direct runner dumps the metrics every time. That's fine.

We likely need an extractor registration set up, to go along with the runner Execute registration, as in the absence of a reliable job management server, there's no common way for runners to return metrics to the users.

**Comment by lostluck** [comment]
Run Go Postcommit

**Pull request merged**